pgxn/neon/walsender_hooks.c

```diff
@@ -0,0 +1,121 @@
+#include "postgres.h"
+#include "fmgr.h"
+#include "access/xlogdefs.h"
+#include "replication/walsender.h"
+#include "access/xlog.h"
+#include "access/xlog_internal.h"
+#include "access/xlogreader.h"
+#include "miscadmin.h"
+#include "utils/wait_event.h"
+
+#include "neon_walreader.h"
+#include "walproposer.h"
+
+static NeonWALReader *wal_reader = NULL;
+extern XLogRecPtr WalSndWaitForWal(XLogRecPtr loc);
+extern bool GetDonorShmem(XLogRecPtr *donor_lsn);
+
+static XLogRecPtr NeonWALReadWaitForWAL(XLogRecPtr loc)
+{
+    while(!NeonWALReaderUpdateDonor(wal_reader))
+        pg_usleep(1000);
+
+    return WalSndWaitForWal(loc);
+}
+
+static int NeonWALPageRead(
+    XLogReaderState *xlogreader,
+    XLogRecPtr targetPagePtr,
+    int reqLen,
+    XLogRecPtr targetRecPtr,
+    char *readBuf)
+{
+    // Wait for flush pointer to advance past our request
+    XLogRecPtr flushptr = NeonWALReadWaitForWAL(targetPagePtr + reqLen);
+    int count;
+    if (flushptr < targetPagePtr + reqLen)
+	return -1;
+
+    // Read at most XLOG_BLCKSZ bytes
+    if (targetPagePtr + XLOG_BLCKSZ <= flushptr)
+	count = XLOG_BLCKSZ;
+    else
+	count = flushptr - targetPagePtr;
+
+    // Sometimes walsender requests non-monotonic sequences of WAL. If that's
+    // the case, we have to reset streaming from remote at the correct position.
+    if(targetPagePtr != NeonWALReaderGetRemLsn(wal_reader))
+    {
+	NeonWALReaderResetRemote(wal_reader);
+    }
+
+    static WaitEventSet *wait_event_set;
+    for(;;)
+    {
+	NeonWALReadResult res = NeonWALRead(
+	    wal_reader,
+	    readBuf,
+	    targetPagePtr,
+	    count,
+	    walprop_pg_get_timeline_id());
+	if(res == NEON_WALREAD_SUCCESS)
+	{
+            // Free event set, since we'll have to recreate it for a new socket later anyway
+	    if (NeonWALReaderEvents(wal_reader) == 0 && wait_event_set != NULL)
+	    {
+		FreeWaitEventSet(wait_event_set);
+		wait_event_set = NULL;
+	    }
+            // Required by xlogreader
+	    xlogreader->seg.ws_tli = NeonWALReaderGetSegment(wal_reader)->ws_tli;
+	    xlogreader->seg.ws_segno = NeonWALReaderGetSegment(wal_reader)->ws_segno;
+	    xlogreader->seg.ws_file = NeonWALReaderGetSegment(wal_reader)->ws_file;
+	    return count;
+	}
+	if(res == NEON_WALREAD_ERROR)
+	{
+	    return -1;
+	}
+        // Res is WOULDBLOCK, so we wait on the socket, recreating event set if necessary
+	{
+	    pgsocket sock = NeonWALReaderSocket(wal_reader);
+	    WaitEvent event;
+	    long timeout_ms = 1000;
+
+	    if (wait_event_set == NULL)
+	    {
+		wait_event_set = CreateWaitEventSet(TopMemoryContext, 3);
+		AddWaitEventToSet(wait_event_set, WL_LATCH_SET, PGINVALID_SOCKET, MyLatch, NULL);
+		AddWaitEventToSet(wait_event_set, WL_EXIT_ON_PM_DEATH, PGINVALID_SOCKET, NULL, NULL);
+		AddWaitEventToSet(wait_event_set, WL_SOCKET_WRITEABLE | WL_SOCKET_READABLE, sock, NULL, NULL);
+	    }
+	    WaitEventSetWait(wait_event_set, timeout_ms, &event, 1, WAIT_EVENT_WAL_SENDER_MAIN);
```

Comment:
We should wait for proper events as indicated by NeonWALReaderEvents. Also note that until connection is established socket might change; hacky  and easy way to work around it is recreate wait event set each time.

Reply:
(it is trivial to observe 100% CPU consumption if you increase 5s sleep in the added test)

Reply:
Made it remake wait_event_set

Reply:
There is still busy loop due to missing
```
        		/* Clear any already-pending wakeups */
		ResetLatch(MyLatch);
``` in the beginning of the loop.

Also, there is WaitLatchOrSocket for this specific use case.

Also, you still have busy loop. Trying to figure out why...

Reply:
Oh sorry, I was wrong about suggestion to remove `count` from NeonWALPageRead yesterday. Indeed, callback can return up to page size of data even if smaller amount is requested, and this is an important optimization. It is weird though that things break when we don't use it...

Reply:
Yeah, turned out xlogreader misbehaves in this case when record begins at the page boundary. In this case XLogDecodeNextRecord calls ReadPageInternal with only SizeOfXLogRecord reqLen which is forwarded to the callback, but that's not enough to read both page header and record header, so ReadPageInternal then fails. Walsender then happily continues to spin %)

Reply:
To summarize, please
- return `count` back
- add ResetLatch(MyLatch);
- add WaitLatchOrSocket

Reply:
Done. Also added a CHECK_FOR_INTERRUPTS for good measure (similar to WalSndWaitForWal).

Reply:
It is not important here because we don't process set latch anyway, but JFYI the convention is to have ResetLatch(MyLatch); immediately after the sleep, not before it to not miss the update.